NetworkingServiceKit/Classes/Networking/FileUploader.swift

```diff
@@ -0,0 +1,134 @@
+//
+//  FileUploader.swift
+//  NetworkingServiceKit
+//
+//  Created by Leandro Perez on 18/03/2021.
+//
+
+import Foundation
+// Approach from https://www.swiftbysundell.com/articles/http-post-and-file-upload-requests-using-urlsession/
+
+/// `FileUploader` uses `URLSessionUploadTask` and `URLSession` directly, it can be used with a `URLSessionConfiguration`, so yo can configure it with a background session
+class FileUploader: NSObject {
+    static var shared = FileUploader(configuration: .backgroundWithPrefixIdentifier())
+    
+    typealias Percentage = Double
+    typealias ProgressHandler = (Percentage) -> Void
+    typealias CompletionHandler = (Result<Void, Error>) -> Void
+    
+    private var progressHandlersByTaskID = [Int : (progress:ProgressHandler, completion: CompletionHandler)]()
+    
+    private let configuration : URLSessionConfiguration
+    
+    private init(configuration: URLSessionConfiguration = .default,
+                  progressHandlersByTaskID: [Int : (progress:ProgressHandler, completion: CompletionHandler)] = [:]) {
+        self.configuration = configuration
+        self.progressHandlersByTaskID = progressHandlersByTaskID
+    }
+    
+    
+    private lazy var urlSession = URLSession(configuration: configuration,
+                                             delegate: self,
+                                             delegateQueue: .main)
+    
+    func upload(data: Data,
+                to targetURL: URL,
+                dispatchQueue : DispatchQueue = .main,
+                progressHandler: @escaping ProgressHandler,
+                completionHandler: @escaping CompletionHandler) {
+        
+        var request = URLRequest(url: targetURL,
+                                 cachePolicy: .reloadIgnoringLocalCacheData)
+        
+        if let imageURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?
+            .appendingPathComponent(targetURL.lastPathComponent)
+            .appendingPathComponent("\(Int.random(in: 1...100000000))") {
```

Comment:
Not much of a concern, but in terms of best practices would it be more consistent to use `UUID().uuidString` for cases like this?

Reply:
Indeed, thanks for that!